internal/ackhandler/sent_packet_history.go

```diff
@@ -2,158 +2,197 @@ package ackhandler
 			return err
 		}
+		if !cont {
+			return nil
+		}
 	}
 	return nil
 }
 
 // FirstOutstanding returns the first outstanding packet.
 func (h *sentPacketHistory) FirstOutstanding() *Packet {
-	el := h.outstandingPacketList.Front()
-	if el == nil {
+	if !h.HasOutstandingPackets() {
 		return nil
 	}
-	return el.Value
+	for _, p := range h.packets {
+		if p != nil && p.outstanding() {
+			return p
+		}
+	}
+	return nil
 }
 
 func (h *sentPacketHistory) Len() int {
-	return len(h.packetMap)
+	return len(h.packets)
 }
 
-func (h *sentPacketHistory) Remove(p protocol.PacketNumber) error {
-	el, ok := h.packetMap[p]
+func (h *sentPacketHistory) Remove(pn protocol.PacketNumber) error {
+	idx, ok := h.getIndex(pn)
 	if !ok {
-		return fmt.Errorf("packet %d not found in sent packet history", p)
+		return fmt.Errorf("packet %d not found in sent packet history", pn)
+	}
+	p := h.packets[idx]
+	if p.outstanding() {
+		h.numOutstanding--
+		if h.numOutstanding < 0 {
+			panic("negative number of outstanding packets")
+		}
+	}
+	h.packets[idx] = nil
+	// clean up all skipped packets directly before this packet number
+	for idx > 0 {
+		idx--
+		p := h.packets[idx]
+		if p == nil || !p.skippedPacket {
+			break
+		}
+		h.packets[idx] = nil
+	}
+	if idx == 0 {
+		h.cleanupStart()
+	}
+	if len(h.packets) > 0 && h.packets[0] == nil {
+		panic("remove failed")
 	}
-	el.List().Remove(el)
-	delete(h.packetMap, p)
 	return nil
 }
 
+// getIndex gets the index of packet p in the packets slice.
+func (h *sentPacketHistory) getIndex(p protocol.PacketNumber) (int, bool) {
+	if len(h.packets) == 0 {
+		return 0, false
+	}
+	first := h.packets[0].PacketNumber
+	if p < first {
+		return 0, false
+	}
+	index := int(p - first)
+	if index > len(h.packets)-1 {
+		return 0, false
+	}
+	return index, true
+}
+
 func (h *sentPacketHistory) HasOutstandingPackets() bool {
-	return h.outstandingPacketList.Len() > 0
+	return h.numOutstanding > 0
 }
 
-func (h *sentPacketHistory) DeletePacketsBefore(t time.Time) {
-	var nextEl *list.Element[*Packet]
-	// we don't iterate outstandingPacketList, as we should not delete outstanding packets.
-	// being outstanding for more than 3*PTO should only happen in the case of drastic RTT changes.
-	for el := h.etcPacketList.Front(); el != nil; el = nextEl {
-		nextEl = el.Next()
-		p := el.Value
+func (h *sentPacketHistory) DeleteBefore(t time.Time) {
+	for i, p := range h.packets {
+		if p == nil {
+			continue
+		}
 		if !p.SendTime.Before(t) {
 			break
 		}
-		delete(h.packetMap, p.PacketNumber)
-		h.etcPacketList.Remove(el)
+		if !p.declaredLost {
+			continue
+		}
+		if p.outstanding() {
+			h.numOutstanding--
+			if h.numOutstanding < 0 {
+				panic("negative number of outstanding packets")
+			}
+		}
+		h.packets[i] = nil
+	}
+	h.cleanupStart()
+}
+
+// delete all nil entries at the beginning of the packets slice
+func (h *sentPacketHistory) cleanupStart() {
+	for i, p := range h.packets {
+		if p != nil {
+			h.packets = h.packets[i:]
```

Comment:
I'm a bit rust on Go slices, is there any chance for a leak here?

Reply:
No, since all the packets up to `i` have been set to `nil`.
At some point we'll need to allocate a new `packets` slice (on `append`), and the old one will be GC'ed. That could be improved by using a ring buffer or similar data structure. We'll tackle this in a future PR.